pypose/optim/solver.py

```diff
@@ -210,3 +213,120 @@ def forward(self, A: Tensor, b: Tensor) -> Tensor:
+
+        >>> # sparse csr example
+        >>> import pypose.optim.solver as ppos
+        >>> crow_indices = torch.tensor([0, 2, 4])
+        >>> col_indices = torch.tensor([0, 1, 0, 1])
+        >>> values = torch.tensor([1, 2, 3, 4], dtype=torch.float)
+        >>> A = torch.sparse_csr_tensor(crow_indices, col_indices, values)
+        >>> A.to_dense()  # visualize
+        tensor([[1., 2.],
+                [3., 4.]])
+        >>> b = torch.tensor([[1.], [2.]])
+        >>> solver = ppos.CG()
+        >>> x = solver(A, b)
+        tensor([-4.4052e-05,  5.0003e-01])
+
+    '''
+    def __init__(self, maxiter=None, tol=1e-5):
+        super().__init__()
+        self.maxiter, self.tol = maxiter, tol
+
+    def forward(self, A: Tensor, b: Tensor, x: Optional[Tensor]=None,
+                M: Optional[Tensor]=None) -> Tensor:
+        '''
+        Args:
+            A (Tensor): the input tensor. It is assumed to be a symmetric
+                positive-definite matrix. Layout is allowed to be COO, CSR, BSR, or dense.
+            b (Tensor): the tensor on the right hand side. Layout could be sparse or dense
+                but is only allowed to be a type that is compatible with the layout of A.
+                In other words, `A @ b` operation must be supported by the layout of A.
+            x (Tensor, optional): the initial guess for the solution. Default: ``None``.
+            M (Tensor, optional): the preconditioner for A. Layout is allowed to be COO,
+                CSR, BSR, or dense. Default: ``None``.
+
+        Return:
+            Tensor: the solved tensor. Layout is the same as the layout of b.
+        '''
+        if A.ndim == b.ndim:
+            b = b.squeeze(-1)
+        else:
+            assert A.ndim == b.ndim + 1, \
+                'The number of dimensions of A and b must differ by 1'
+        if x is None:
+            x = torch.zeros_like(b)
+        bnrm2 = torch.linalg.norm(b, dim=-1)
+        if (bnrm2 == 0).all():
+            return b
+        atol = self.tol * bnrm2
+        n = b.shape[-1]
+
+        if self.maxiter is None:
+            maxiter = n*10
+        else:
+            maxiter = self.maxiter
+        r = b - bmv(A, x) if x.any() else b.clone()
+        rho_prev, p = None, None
+
+        for iteration in range(maxiter):
+            if (torch.linalg.norm(r, dim=-1) < atol).all():
+                return x
+
+            z = bmv(M, r) if M is not None else r
+            rho_cur = vecdot(r, z)
+            if iteration > 0:
+                beta = rho_cur / rho_prev
+                p =  p * beta.unsqueeze(-1) + z
+            else:  # First spin
+                p = torch.empty_like(r)
+                p[:] = z[:]
+
+            q = bmv(A, p)
+            alpha = rho_cur / vecdot(p, q)
+            x += alpha.unsqueeze(-1)*p
+            r -= alpha.unsqueeze(-1)*q
+            rho_prev = rho_cur
+
+        else:  # for loop exhausted
+            # Return incomplete progress
+            return x
```

Comment:
how about 
```
        if self.maxiter is None:
            maxiter = n * 10
        else:
            maxiter = self.maxiter
        r = b - bmv(A, x) if x.any() else b.clone()
        p = torch.empty_like(r)
        rho_prev, p[:]= None, z[:]

        for iteration in range(maxiter):
            if (torch.linalg.norm(r, dim=-1) < atol).all():
                return x

            z = bmv(M, r) if M is not None else r
            rho_cur = vecdot(r, z)
            beta = rho_cur / rho_prev
            p = p * beta.unsqueeze(-1) + z
            q = bmv(A, p)
            alpha = rho_cur / vecdot(p, q)
            x += alpha.unsqueeze(-1)*p
            r -= alpha.unsqueeze(-1)*q
            rho_prev = rho_cur

        return x
```

Reply:
z has to be defined before it is used in `rho_prev, p[:]= None, z[:]`